File: 93.restore-ip-addresses.java
/*
 * @lc app=leetcode id=93 lang=java
 *
 * [93] Restore IP Addresses
 */

// @lc code=start
class Solution {

    List<String> res = new LinkedList<>();

    public List<String> restoreIpAddresses(String s) {
        if (s != null && s.length() >= 4) resolve("", s, 1);
        return res;
    }

    private void resolve(String cur, String s, int segment) {
        if (segment == 5) {
            if (s.length() == 0) res.add(cur.substring(1));
            return;
        }
        if (s.length() >= 1) {
            resolve(cur + "." + s.substring(0,1), s.substring(1), segment+1);
            if (s.charAt(0) == '0') return;
        }
        if (s.length() >= 2) {
            resolve(cur + "." + s.substring(0,2), s.substring(2), segment+1);
        }
        if (s.length() >= 3) {
            if (Integer.valueOf(s.substring(0,3)) <= 255) {
                resolve(cur + "." + s.substring(0,3), s.substring(3), segment+1);
            }
        }

    }
}
// @lc code=end

File: go/93.restore-ip-addresses.go
/*
 * @lc app=leetcode id=93 lang=golang
 *
 * [93] Restore IP Addresses
 */

// @lc code=start
func restoreIpAddresses(s string) []string {
    return tryseperate(s, 4)
}

func tryseperate(s string, ndots int) []string {
	res := []string{}
	for i := 1; i <= len(s) && i <= 3; i++ {
		if i > 1 && s[:1] == "0" {
			break;
		}
		if val, _ := strconv.Atoi(s[:i]); val <= 255 {
			if ndots == 1 {
				if i == len(s) {
					res = append(res, s)
				}
			} else {
				offsets := tryseperate(s[i:], ndots-1)
				for _, offset := range offsets {
					res = append(res, s[:i]+"."+offset)
				}
			}
		}
	}
	return res
}
// @lc code=end

